win32/src/file_win32.c

```diff
@@ -60,4 +40,448 @@ static void on_file_io_complete_win32(
+                    result->ptp_cleanup_group = CreateThreadpoolCleanupGroup();
+
+                    if (result->ptp_cleanup_group == NULL)
+                    {
+                        /*Codes_SRS_FILE_43_034: [ If there are any failures, file_create shall fail and return NULL. ]*/
+                        LogLastError("Failure in CreateThreadpoolCleanupGroup, full_file_name=%s", full_file_name);
+                        succeeded = false;
+                    }
+                    else
+                    {
+                        /*Codes_SRS_FILE_WIN32_43_007: [ file_create shall register the cleanup group with the threadpool environment by calling SetThreadpoolCallbackCleanupGroup.]*/
+                        SetThreadpoolCallbackCleanupGroup(&result->cbe, result->ptp_cleanup_group, on_close_threadpool_group_member);
+
+                        /*Codes_SRS_FILE_WIN32_43_033: [ file_create shall create a threadpool io with the allocated FILE_HANDLE and on_file_io_complete_win32 as a callback by calling CreateThreadpoolIo]*/
+                        result->ptp_io = CreateThreadpoolIo(result->h_file, on_file_io_complete_win32, NULL, &result->cbe);
+                        
+                        if (result->ptp_io == NULL)
+                        {
+                            /*Codes_SRS_FILE_43_034: [ If there are any failures, file_create shall fail and return NULL. ]*/
+                            /*Codes_SRS_FILE_WIN32_43_008: [ If there are any failures, file_create shall return NULL. ]*/
+                            LogLastError("Failure in CreateThreadpoolIo, full_file_name=%s", full_file_name);
+                            succeeded = false;
+                        }
+                        else
+                        {
+                            /*Codes_SRS_FILE_WIN32_43_009: [ file_create shall succeed and return a non - NULL value.]*/
+                            succeeded = true;
+                            result->user_report_fault_callback = user_report_fault_callback;
+                            result->user_report_fault_context = user_report_fault_context;
+                        }
+                        
+                        if (!succeeded)
+                        {
+                            CloseThreadpoolCleanupGroup(result->ptp_cleanup_group);
+                        }
+                    }
+                    if (!succeeded)
+                    {
+                        DestroyThreadpoolEnvironment(&result->cbe);
+                    }
+                }
+            }
+            /*Codes_SRS_FILE_43_034: [ If there are any failures, file_create shall fail and return NULL. ]*/
+            if (!succeeded)
+            {
+                if (!CloseHandle(result->h_file))
+                {
+                    LogLastError("Failure in CloseHandle, full_file_name=%s", full_file_name);
+                }

```

Comment:
This belongs up between lines 168-169 because it should happen only when CreateFileA succeeds